Add several memory helpers (required for speed up of string search).

Rot127 opened the pull request:
<!-- Filling this template is mandatory -->

**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository.
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style).
- [x] I've documented every `RZ_API` function and struct this PR changes.
- [x] I've added tests that prove my changes are effective (required for changes to `RZ_API`).
- [ ] I've updated the [Rizin book](https://github.com/rizinorg/book) with the relevant information (if needed).

**Detailed description**

<!-- Explain the **details** for making this change. Is a new feature implemented? What existing problem does the pull request solve? How does the pull request solve these issues? Please provide enough information so that others can review your pull request. -->

Adds a few more memory and bit related helpers.
I'll use them for the speed up of the string search of unaligned memory and foreign endianness encodings.

Functions added:

- Count trailing zeros
- Check pointer alignment
- Memcpy with offset into src.
- Swap byte blocks of 2 and 4 bytes.

**Test plan**

<!-- What steps should the reviewer take to test your pull request? Demonstrate the code is solid. Example: The exact commands you ran and their output, screenshots/videos. This is your time to re-check that everything works and that you covered all the edge cases -->

Added

**Closing issues**

<!-- put "closes #XXXX" in your comment to auto-close the issue that your PR fixes (if any). -->

...

Rot127 committed:
Add function to memcpy memory with offset.

Rot127 committed:
Add count trailing zeros function.

Rot127 committed:
Add function to determine pointer alignment.

Rot127 committed:
Add helper to swap sequences of 2 bytes.

Rot127 committed:
Add helper to swap sequences of 4 bytes.

Rot127 committed:
Print pointer in warning

Rot127 committed:
Fix tests

Rot127 commented:
Apparently do most compilers produce code with unaligned data.
Will fix it tomorrow.

Rot127 committed:
Use uintptr_t type to prevent undefined behavior.

Rot127 committed:
Remove alignment requirement and rely on malloc promises.

Rot127 committed:
Add in place byte swap functions.

notxvilka commented on librz/include/rz_util/rz_bits.h:
```diff
@@ -33,6 +33,45 @@ DEFINE_COUNT_ONES(ut32);
 DEFINE_COUNT_ONES(ut16);
 DEFINE_COUNT_ONES(ut8);
 
+static inline size_t rz_bits_trailing_zeros(ut64 v) {
```
Add Doxygen

notxvilka replied:
If I remember correctly, compilers should have `__builtin_ctz()` for that. Please check if could be used instead.

notxvilka replied:
See https://gcc.gnu.org/onlinedocs/gcc/Bit-Operation-Builtins.html
Please detect it via Meson like other builtins

notxvilka commented on librz/util/mem.c:
```diff
@@ -331,3 +331,251 @@ RZ_API void rz_mem_memzero(void *dst, size_t l) {
+}
+
+/**
+ * \brief Swaps the bytes in 2 byte blocks from the given buffer and returns
```
Are you aware of `rz_swap_*()` functions in the `librz/include/rz_endian.h`?

Rot127 replied:
Ah, damn. I actually wasn't. Must have overlooked them. Will use them here.

Rot127 committed:
Move the endianness swapping into util/rz_endian

Rot127 committed:
Use GNU builtin __builtin_ctzll/clzll.

Also fixes the include guards of __builtin_clzll
and handles the undefined case of x == 0 of it.

Rot127 committed:
Add doxygen

notxvilka commented on meson.build:
```diff
@@ -439,6 +439,7 @@ foreach it : ccs
   it_userconf.set10('HAVE_ZLIB', get_option('use_zlib'))
   it_userconf.set10('SUPPORTS_PCRE2_JIT', pcre2_jit_supported)
   it_userconf.set('N_THREAD_LIMIT', get_option('n_thread_limit'))
+  it_userconf.set10('HAVE___BUILTIN_CTZLL', cc.has_function('__builtin_ctzll'))
```
Why not put together with BUILTIN_CLZLL?

Rot127 replied:
I don't know why. But the `HAVE___BUILTIN_CTZLL` value is simply not handled by Muon. Meson gives a warning as well, but doesn't fail. It just generates
`#define HAVE___BUILTIN_CTZLL ` (without `1` or `0`).

Rot127 commented:
Removed the merge-when-green` label. Please let us merge the CI fixes first and then I can rebase. Just in case...

notxvilka reviewed (changes requested):
It breaks MSVC builds:
```
"link"  /MACHINE:x64 /OUT:librz/util/sdb_gen.exe librz/util/sdb_gen.exe.p/sdb_src_main.c.obj "/release" "/nologo" "/OPT:REF" "librz/util/rz_util.lib" "ws2_32.lib" "wininet.lib" "psapi.lib" "/SUBSYSTEM:CONSOLE" "kernel32.lib" "user32.lib" "gdi32.lib" "winspool.lib" "shell32.lib" "ole32.lib" "oleaut32.lib" "uuid.lib" "comdlg32.lib" "advapi32.lib"
   Creating library librz\util\sdb_gen.lib and object librz\util\sdb_gen.exp
rz_util.lib(mem.c.obj) : error LNK2019: unresolved external symbol __builtin_ctzll referenced in function rz_bits_trailing_zeros
librz\util\sdb_gen.exe : fatal error LNK1120: 1 unresolved externals
[1770/2404] Linking static target librz/hash/rz_hash.lib
[1771/2404] Compiling C object librz/main/rz_main.lib.p/rz-ax.c.obj
ninja: build stopped: subcommand failed.
```

wargio commented on librz/include/rz_types_base.h:
```diff
@@ -27,6 +27,7 @@
 #define st16  short
 #define ut8   unsigned char
 #define st8   signed char
+#define utptr uintptr_t
```
not super sure this is standard, because the `_t` suffix is usually a typedef.

wargio replied:
Oh, it's part of in C99

wargio replied:
maybe we might need to patch this by also checking that `uintptr_t` exists otherwise use `size_t`

notxvilka replied:
It is: https://en.cppreference.com/w/c/types/integer.html

notxvilka replied:
No need, it's a strong C99 requirement and our minimally supported version.

Rot127 replied:
`uintptr_t` is the only type which is guaranteed to have a conversion defined between pointer <-> integer. `size_t` isn't to my knowledge.
We could just copy the `#ifdef` from the std headers, but I think `uintptr_t` pretty common since it was already in C99. So if this fails, other parts of our code probably fail as well.

wargio replied:
since c99 :dango:

notxvilka replied:
No need. If we ever meet compiler that doesn't know it - we will think about it. Now, everything is at least C99. Not an issue.
